app/feature/feature-terminate-insurance/src/main/kotlin/com/hedvig/android/feature/terminateinsurance/step/choose/ChooseInsuranceToTerminateViewModel.kt

```diff
@@ -0,0 +1,170 @@
+import com.hedvig.android.data.termination.data.TerminatableInsurance
+import com.hedvig.android.feature.terminateinsurance.InsuranceId
+import com.hedvig.android.feature.terminateinsurance.data.TerminateInsuranceRepository
+import com.hedvig.android.feature.terminateinsurance.data.TerminateInsuranceStep
+import com.hedvig.android.logger.LogPriority
+import com.hedvig.android.logger.logcat
+import com.hedvig.android.molecule.android.MoleculeViewModel
+import com.hedvig.android.molecule.public.MoleculePresenter
+import com.hedvig.android.molecule.public.MoleculePresenterScope
+
+internal class ChooseInsuranceToTerminateViewModel(
+  insuranceId: String?,
+  getTerminatableContractsUseCase: GetTerminatableContractsUseCase,
+  terminateInsuranceRepository: TerminateInsuranceRepository,
+) : MoleculeViewModel<ChooseInsuranceToTerminateEvent, ChooseInsuranceToTerminateStepUiState>(
+    initialState = ChooseInsuranceToTerminateStepUiState.Loading,
+    presenter = ChooseInsuranceToTerminatePresenter(
+      insuranceId = insuranceId,
+      getTerminatableContractsUseCase = getTerminatableContractsUseCase,
+      terminateInsuranceRepository = terminateInsuranceRepository,
+    ),
+  )
+
+private class ChooseInsuranceToTerminatePresenter(
+  private val insuranceId: String?,
+  private val getTerminatableContractsUseCase: GetTerminatableContractsUseCase,
+  private val terminateInsuranceRepository: TerminateInsuranceRepository,
+) : MoleculePresenter<ChooseInsuranceToTerminateEvent, ChooseInsuranceToTerminateStepUiState> {
+  @Composable
+  override fun MoleculePresenterScope<ChooseInsuranceToTerminateEvent>.present(
+    lastState: ChooseInsuranceToTerminateStepUiState,
+  ): ChooseInsuranceToTerminateStepUiState {
+    var loadIteration by remember { mutableIntStateOf(0) }
+    var currentSelectedId by remember {
+      val initialSelected = if (lastState is ChooseInsuranceToTerminateStepUiState.Success) {
+        lastState.selectedInsurance?.id
+      } else {
+        insuranceId
+      }
+      mutableStateOf(initialSelected)
+    }
+    var terminationStep: TerminateInsuranceStep? by remember { mutableStateOf(null) }
+    var currentPartialState by remember {
+      mutableStateOf(lastState.toPartialState())
+    }
+
+    CollectEvents { event ->
+      when (event) {
+        ChooseInsuranceToTerminateEvent.RetryLoadData -> {
+          loadIteration++
+        }
+
+        is ChooseInsuranceToTerminateEvent.SelectInsurance -> {
+          currentSelectedId = event.insurance.id
+          if (currentPartialState is PartialUiState.Success) {
+            currentPartialState =
+              (currentPartialState as PartialUiState.Success).copy(selectedInsurance = event.insurance)
+          }
+        }
+      }
+    }
+
+    LaunchedEffect(currentSelectedId) {
+      val id = currentSelectedId
+      if (id != null) {
+        terminationStep = terminateInsuranceRepository.startTerminationFlow(InsuranceId(id)).getOrNull()
+      }
+    }
```

Comment:
Aha I see, you fetch the next step here on each change of the currentSelectedId (which is also what comes as a parameter here if we want to have something pre-selected).

Two thoughts here:
* If the user goes back and forth between options, on each one of them we do a request to start the termination flow as a side effect. This comes at the benefit that then when you click the button you already know where to go instead of loading.
* If this fails, so when we do `getOrNull()` it is null, there does not seem to be any way for the user to retry this besides picking another option and coming back to this again.
* This also makes it so that the button to move forward does not show as enabled as soon as you click the button, but this needs to run first.

Do you think we can kinda revert the order in which things are happening here to avoid the above issues?
So when the user picks one of the options (or one is pre-selected even), we just keep that selection locally in the state but without triggering any network request.
Then when the button is pressed (and it should be able to be pressed immediately as long as there is a selection in the first place) then we do the network request, and we show a loading spinner inside the button.
This way if this fails, the user can actually click the button again to try again in some way. We do not have design for what happens if that fails of course right now, but for now it can even be a Toast till we discuss this with design.

In general though this flip of the concerns needs to happen here so that we do not end up in these unrecoverable scenarios if `startTerminationFlow` fails.

That way I think you'll also be able to get rid of the partial state and keep one UiState. Probably, you can figure out if its still not possible yet as you do this

Reply:
Fixed in the last commit:)